benchmarks/microbenchmarks/utils.py

```diff
@@ -0,0 +1,380 @@
+                filter_fn=ffn_or_attn_only,
+            )
+    if "fp6" in quantization:
+        quantize_(model, FPXWeightOnlyConfig(3, 2))
+    elif "embed-int8wo" in quantization:
+        quantize_(
+            model,
+            Int8WeightOnlyConfig(group_size=64),
+            filter_fn=lambda x, *args: isinstance(x, torch.nn.Embedding),
+        )
+    elif "uintx" in quantization:
+        # uintx-nbits-group_size, e.g. "uintx-2-64"
+        if "hqq" in quantization:
+            # uintx-nbits-group_size-hqq
+            use_hqq = True
+        else:
+            use_hqq = False
+        _quant_args = quantization.split("-")
+        nbits = int(_quant_args[1])
+        assert nbits >= 1 and nbits <= 8, "nbits must be 1 to 8"
+        _NBITS_TO_DTYPE = {
+            1: torch.uint1,
+            2: torch.uint2,
+            3: torch.uint3,
+            4: torch.uint4,
+            5: torch.uint5,
+            6: torch.uint6,
+            7: torch.uint7,
+            8: torch.uint8,
+        }
+        dtype = _NBITS_TO_DTYPE[nbits]
+        group_size = int(_quant_args[2])
+        quantize_(model, UIntXWeightOnlyConfig(dtype, group_size, use_hqq=use_hqq))
+    elif "int8_dynamic_activation_intx_weight" in quantization:
+        from torchao.experimental.quant_api import (
+            int8_dynamic_activation_intx_weight,
+        )
+        from torchao.quantization.granularity import PerGroup
+
+        assert (
+            precision == torch.float32
+        ), "int8_dynamic_activation_intx_weight requires using precision=torch.float32"
+
+        # Quantize model
+        _quant_args = quantization.split("-")
+        weight_dtype = getattr(torch, f"int{_quant_args[1]}")
+        granularity = PerGroup(int(_quant_args[2]))
+        has_weight_zeros = bool(_quant_args[3])
+        quantize_(
+            model,
+            int8_dynamic_activation_intx_weight(
+                weight_dtype=weight_dtype,
+                granularity=granularity,
+                has_weight_zeros=has_weight_zeros,
+            ),
+        )
+    elif "float8wo" in quantization:
+        quantize_(model, Float8WeightOnlyConfig())
+    elif "float8dq" in quantization:
+        granularity = str(quantization.split("-")[-1])
+        if granularity == "tensor":
+            granularity = PerTensor()
+        elif granularity == "row":
+            granularity = PerRow()
+        else:
+            granularity = PerTensor()
+        quantize_(
+            model, Float8DynamicActivationFloat8WeightConfig(granularity=granularity)
+        )
+    else:
+        if not TORCH_VERSION_AT_LEAST_2_5:
+            unwrap_tensor_subclass(model)
+    return model
+
+
+# Function to benchmark model evaluation - e2e eval run
+def benchmark_model_inference_in_microseconds(model, input_data):
+    # Returns model run time in seconds
+    if torch.cuda.is_available():
+        torch.cuda.synchronize()
+
+    # warm up
+    for _ in range(2):
```

Comment:
for example the warmup without torch.no_grad vs run with torch.no_grad is going to create diverging behaviors.

also, in practice if the first run of the model is where compile actually occurs, you're not really warming things up.